### PR TITLE
Add support to create additional bindings

### PR DESCRIPTION
**Pull request by SheGe**
Hey,
I've added support to create additional binding types.
Code is not formatted in same style as rest. Sorry for that, but I cannot find description for contributors.

Work summary in commit comments.

### CODE REVIEW

**Comment by SharpeRAD** [comment]
Looks good, if you could document the public methods / properties / constructors you've added as I've been trying to complete the documentation in this project for ages but it keeps growing so I keep putting it off :anguished:

See [BaseManager](https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Manager/Base/BaseManager.cs) as an example.

**Comment by SheGe** [comment]
Hey,

I've updated documentation and some logic around the bindings.
The most important change for user is that declaration of AddSiteBinding methods were changed (siteName parameter has been added):

`        public static void AddSiteBinding(this ICakeContext context, string siteName, IBindingSettings settings)`
`        public static void AddSiteBinding(this ICakeContext context, string server, string siteName, IBindingSettings settings)`

**Change will make huge impact on the end users who used these methods before.**

File: src/IIS/Settings/Bindings/CustomBindingSettings.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        /// <inheritdoc />
+        public virtual string BindingInformation
+        {
+            get
+            {
+                return string.Format(@"{0}:{1}:{2}", IpAddress, Port, HostName);
+            }
+        }
+
+        ICustomBindingSettings ICustomBindingSettings.Port(int port)
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/CustomBindingSettings.cs]: Prefix function names with "Set" and change them to generic extension methods around an IBindingSettings interface. [Use this as an example](https://github.com/SharpeRAD/Cake.AWS.S3/blob/master/src/S3/Extensions/Settings/S3SettingsExtensions.cs)

File: src/IIS/Settings/Bindings/FluentAPI/IBindings.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+﻿namespace Cake.IIS.Settings.Bindings.FluentAPI
+{
+    public interface IBindings
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/FluentAPI/IBindings.cs]: Why use an interface when your effectively trying to mimic an Enum of classes? Drop the interface and just use a static class to expose the individual implementations

File: src/IIS/Settings/Bindings/FluentAPI/ICustomBindingSettings.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+﻿namespace Cake.IIS.Settings.Bindings.FluentAPI
+{
+    /// <summary>
+    /// Fluent interface to configure any type of binding.
+    /// </summary>
+    public interface ICustomBindingSettings
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/FluentAPI/ICustomBindingSettings.cs]: These Interfaces should declare properties, not methods.

File: src/IIS/Settings/Bindings/FluentAPI/ICustomBindingSettings.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        /// Sets certificate hash.
+        /// </summary>
+        /// <param name="hashSet">The certificate hash.</param>
+        ICustomBindingSettings CertificateHash(byte[] hashSet);
+
+        /// <summary>
+        /// Gets instance of <see cref="CustomBindingSettings"/>.
+        /// </summary>
+        /// <returns>Fluently created instance of type <see cref="CustomBindingSettings"/>.</returns>
+        CustomBindingSettings Instance { get; }
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/FluentAPI/ICustomBindingSettings.cs]: This would not be required if the Interface declared properties instead of methods.

File: src/IIS/Settings/Bindings/FluentAPI/IHttpBindingSettings.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿namespace Cake.IIS.Settings.Bindings.FluentAPI
+{
+    /// <summary>
+    /// Fluent interface to configure http binding.
+    /// </summary>
+    public interface IHttpBindingSettings
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/FluentAPI/IHttpBindingSettings.cs]: These interfaces should share properties and inherit from one another. eg: IBindingSettings, IHostBindingSettings, ICertificateBindingSettings

File: src/IIS/Settings/Bindings/IISBindings.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using Cake.IIS.Settings.Bindings.FluentAPI;
+
+namespace Cake.IIS.Settings.Bindings
+{
+    internal class IISBindings : IBindings
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/IISBindings.cs]: This should be a static class, exposing the implementations from static functions

File: src/IIS/Settings/Bindings/IISBindings.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        private readonly Action<IBindingSettings> _bindingCreationRelay;
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/IISBindings.cs]: Would not be required if your interfaces exposed properties, not methods

File: src/IIS/Settings/Bindings/MsmqFormatNameBindingSettings.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
+﻿using Cake.IIS.Settings.Bindings.FluentAPI;
+
+namespace Cake.IIS.Settings.Bindings
+{
+    /// <summary>
+    /// Class to represent and configure msmq.formatname binding.
+    /// </summary>
+    public sealed class MsmqFormatNameBindingSettings : IMsmqFormatNameBindingSettings, IBindingSettings
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/MsmqFormatNameBindingSettings.cs]: Why don't the individual bindings inherit from a base class since they share the majority of their properties?

File: src/IIS/Settings/SiteSettings.cs [hunk 9]
```diff
@@ -23,34 +33,107 @@ public SiteSettings()
...
+            [Obsolete("Use ChangeBinding or ChangeBindingTo method instead.")]
+            public BindingProtocol BindingProtocol
+            {
+                get { return bindingSettings.BindingProtocol; }
+                set { bindingSettings.BindingProtocol = value; }
+            }
+
+        #endregion
+
+        #region Methods (2)
```
Comment (SharpeRAD) [inline src/IIS/Settings/SiteSettings.cs]: Neither of these methods would be required if the binding interfaces exposed properties

File: src/IIS/Aliases/SiteAliases.cs [hunk 10]
```diff
@@ -17,6 +20,17 @@ namespace Cake.IIS
     public static class SiteAliases
     {
         /// <summary>
+        /// Gets pre-defined bindings
+        /// </summary>
+        /// <param name="context">The context</param>
+        /// <returns>Predefined bindings for IIS</returns>
+        [CakePropertyAlias]
+        public static IBindings IISBindings(this ICakeContext context)
```
Comment (SharpeRAD) [inline src/IIS/Aliases/SiteAliases.cs]: Wouldn't be required if IISBindings was a static class

File: src/IIS/Settings/Bindings/FluentAPI/IMsmqFormatNameBindingSettings.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Cake.IIS.Settings.Bindings.FluentAPI
+{
+    /// <summary>
+    /// Fluent interface to configure msmq.formatname binding.
+    /// </summary>
+    public interface IMsmqFormatNameBindingSettings
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/FluentAPI/IMsmqFormatNameBindingSettings.cs]: The individual implementations don't need their own interfaces.

File: src/IIS/Settings/Bindings/IISBindings.cs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        /// <inheritdoc />
+        public ICustomBindingSettings CustomBinding(BindingProtocol protocol)
+        {
+            var customBindingSettings = new CustomBindingSettings(protocol);
+            NotifyAboutCreation(customBindingSettings);
+            return customBindingSettings;
+        }
+
+
+        private void NotifyAboutCreation(IBindingSettings bindingSettings)
```
Comment (SharpeRAD) [inline src/IIS/Settings/Bindings/IISBindings.cs]: Shouldn't be required

File: src/IIS/Settings/FtpsiteSettings.cs [hunk 13]
```diff
@@ -6,8 +6,7 @@ public class FtpsiteSettings : SiteSettings
             public FtpsiteSettings()
                 : base()
             {
-                this.BindingProtocol               = BindingProtocol.Ftp;
-                this.Port                          = 21;
+                ChangeBindingTo().Ftp();
```
Comment (SharpeRAD) [inline src/IIS/Settings/FtpsiteSettings.cs]: This could be as simple as this.Binding = IISBindings.Ftp() if you achanged the interfaces to expose properties and made IISBindings static

File: src/IIS/Settings/SiteSettings.cs [hunk 14]
```diff
@@ -1,13 +1,23 @@
...
 #endregion
 
 
 
 namespace Cake.IIS
 {
-    public abstract class SiteSettings : BindingSettings, IDirectorySettings
+    public abstract class SiteSettings : IDirectorySettings
     {
+        #region Fields (1)
```
Comment (SharpeRAD) [inline src/IIS/Settings/SiteSettings.cs]: Why is this here? Your effectively allowing people to update something via obsolete methods that isn't actually used by the Managers.

**Comment by SheGe** [comment]
Thanks! Over Engineering. I'll update the code in few days.

File: src/IIS/Settings/SiteSettings.cs (hunk 14 above)
Comment (SheGe) [inline src/IIS/Settings/SiteSettings.cs]: This is a bug. I forgot about updating Managers code.

File: src/IIS/Settings/Bindings/FluentAPI/IBindings.cs (hunk 2 above)
Comment (SheGe) [inline src/IIS/Settings/Bindings/FluentAPI/IBindings.cs]: Interface exposed Fluent API only. It will be removed with extension methods implementation.
